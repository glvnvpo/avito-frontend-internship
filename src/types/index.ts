// @flow

export type ID = number | string;

export type URL = string;

export type Story = {
    id: ID;
    title: string;
    by: string;
    time: number | string;
    score: number | string;
    kids?: Array<number | string>;
    text?: string;
    url?: string;
    descendants?: number | string;
}

export type ChildComment = {
    id: number | string;
    by: string;
    text: string;
    time: number | string;
    kids?: Array<number | string>;
}

export type Comment = {
    id: number | string;
    by: string;
    text: string;
    time: number | string;
    kids?: Array<number | string>;
    children?: Array<ChildComment>;
    showChildComment?: boolean;
    isLoadingChildren?: boolean;
}