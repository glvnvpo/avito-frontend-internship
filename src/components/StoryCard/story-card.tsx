// @flow

import React, {FC, HTMLAttributes} from 'react';
import {useNavigate} from 'react-router-dom';
import {Card} from 'antd';
import {isEmpty} from 'lodash';
import './styles.scss';
import {Story} from '../../types';
import {getDateFromTimestamp} from '../../helpers/get-date-from-timestamp';

export enum Fields {
    ALL= 'all',
    URL = 'url',
    COMMENTS_COUNT = 'comments-count',
	SCORE = 'score'
}

type Props = {
    story: Story | undefined;
    isLoading?: boolean;
    asLink?: boolean;
    to?: string;
	extraFieldsToShow?: Array<Fields>;
}

export const StoryCard: FC<Props & HTMLAttributes<any>> = ({story, isLoading=false, asLink=false,
	to, extraFieldsToShow=[Fields.ALL], ...rest}) => {
    
	let {title, score, by, time, url, descendants} = story || {};

	const navigate = useNavigate();

	const shouldShowExtraField = (key: Fields): boolean => {
		if (extraFieldsToShow.includes(Fields.ALL)) {
			return true;
		}

		return extraFieldsToShow.includes(key);
	};
	
	const navigateTo = (path: string | undefined): void => {
		if (path) {
			navigate(path);
		}
	};

	return (
		<Card className={`story color-grey3 ${asLink && 'link'}`}
			  onClick={() => navigateTo(to)} {...rest} loading={isLoading}>

			{
				!isEmpty(story) ? 
					<>
						<span className='title color-orange bold mb-10'>{title}</span>
						<span>
							{ shouldShowExtraField(Fields.SCORE) && <>{score}&nbsp;point{(score && score>1) &&'s'}&nbsp;| </> }
							<span className='user bold'>{by}</span>&nbsp;| {getDateFromTimestamp(time)}</span>

						{
							shouldShowExtraField(Fields.URL) &&
							<span className='url'>{url ? <a href={url} target='_blank' rel='noreferrer'>Visit source</a> : 'No source link available'}</span>
						}

						{
							shouldShowExtraField(Fields.COMMENTS_COUNT) &&
							<span className='comments-count'>Comments count: {descendants}</span>
						}
					</>
					: <span className='error bold'>Some troubles in loading story</span>
			}
		</Card>

	);
};