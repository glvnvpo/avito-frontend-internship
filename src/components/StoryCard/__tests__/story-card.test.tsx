import React from 'react';
import {mount} from 'enzyme';
import {BrowserRouter} from 'react-router-dom';
import {Fields, StoryCard} from '../story-card';
import {Story} from '../../../types';

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockUseNavigate
}));

type Props = {
	story: Story | undefined;
	isLoading?: boolean;
	asLink?: boolean;
	to?: string;
	extraFieldsToShow?: Array<Fields>;
}

describe('StoryCard', () => {
	let wrapper;

	const renderComponent = ({story, ...rest}: Props) =>
		mount(<BrowserRouter>
			<StoryCard story={story} {...rest}/>
		</BrowserRouter>);


	const story = {
		id: 11,
		title: 'someTitle',
		by: 'coolUser',
		time: 1666689031,
		score: 3,
		url: 'https://reactjs.org/',
		descendants: 10
	};

	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should render correct story card', () => {
		wrapper = renderComponent({story: story});
		expect(wrapper).toMatchSnapshot();
	});

	it('should call useNavigate hook', () => {
		wrapper = renderComponent({story: story, asLink: true, to:'/somePath'});
		wrapper.find('.story>.link').simulate('click');
		expect(mockUseNavigate).toHaveBeenCalled();
	});

	it('should not call useNavigate hook when path is not defined', () => {
		wrapper = renderComponent({story: story, asLink: true});
		wrapper.find('.story>.link').simulate('click');
		expect(mockUseNavigate).not.toHaveBeenCalled();
	});

	it('should render extra field', () => {
		wrapper = renderComponent({story: story, extraFieldsToShow: [Fields.COMMENTS_COUNT]});
		expect(wrapper.find('.url')).toHaveLength(0);
		expect(wrapper.find('.comments-count')).toHaveLength(1);
	});

	it('should render error when url is undefined', () => {
		const newStory = {
			...story,
			url: undefined
		};
		wrapper = renderComponent({story: newStory, extraFieldsToShow: [Fields.URL]});
		expect(wrapper.find('.url').text()).toBe('No source link available');
	});

	it('should render when story is undefined', () => {
		wrapper = renderComponent({story: undefined});
		expect(wrapper.text()).toBe('Some troubles in loading story');
	});
});