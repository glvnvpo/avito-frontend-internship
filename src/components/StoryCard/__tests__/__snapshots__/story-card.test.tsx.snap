// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryCard should render correct story card 1`] = `
<BrowserRouter>
  <Router
    location={
      {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigationType="POP"
    navigator={
      {
        "action": "POP",
        "createHref": [Function],
        "go": [Function],
        "listen": [Function],
        "location": {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <StoryCard
      story={
        {
          "by": "coolUser",
          "descendants": 10,
          "id": 11,
          "score": 3,
          "time": 1666689031,
          "title": "someTitle",
          "url": "https://reactjs.org/",
        }
      }
    >
      <ForwardRef
        className="story color-grey3 false"
        loading={false}
        onClick={[Function]}
      >
        <div
          className="ant-card ant-card-bordered story color-grey3 false"
          onClick={[Function]}
        >
          <div
            className="ant-card-body"
            style={{}}
          >
            <span
              className="title color-orange bold mb-10"
            >
              someTitle
            </span>
            <span>
              3
               point
              s
               | 
              <span
                className="user bold"
              >
                coolUser
              </span>
               | 
              25.10.2022
            </span>
            <span
              className="url"
            >
              <a
                href="https://reactjs.org/"
                rel="noreferrer"
                target="_blank"
              >
                Visit source
              </a>
            </span>
            <span
              className="comments-count"
            >
              Comments count: 
              10
            </span>
          </div>
        </div>
      </ForwardRef>
    </StoryCard>
  </Router>
</BrowserRouter>
`;
